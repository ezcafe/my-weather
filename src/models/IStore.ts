import { RouterState } from 'connected-react-router';
import IErrorState from '../stores/error/models/IErrorState';
import ILocationsState from '../stores/locations/models/ILocationsState';
import IRequestingState from '../stores/requesting/models/IRequestingState';
import IToastsState from '../stores/toasts/models/IToastsState';

export default interface IStore {
    readonly error: IErrorState;
    readonly requesting: IRequestingState;
    readonly router: RouterState;
    readonly locations: ILocationsState;
    readonly toasts: IToastsState;
}
