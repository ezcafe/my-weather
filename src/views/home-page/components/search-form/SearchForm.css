.search-form {
    margin-top: 10px;
}

.input-group-prepend {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;
}
