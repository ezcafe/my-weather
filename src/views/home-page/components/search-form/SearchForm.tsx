import { useCallback, useMemo, useState } from 'react';
import { InputGroup, Form } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import Select, { OptionTypeBase } from 'react-select';
import { Dispatch } from 'redux';
import {
    REQUEST_LOCATIONS,
    requestLocationsAction,
    requestLocationWeatherAction,
} from '../../../../stores/locations/LocationsAction';
import { locationsSelector } from '../../../../stores/locations/LocationsSelector';
import { selectRequesting } from '../../../../stores/requesting/RequestingSelector';
import LocationModel from '../../../../stores/locations/models/LocationModel';

import './SearchForm.css';

const customStyles = {
    container: () => ({
        width: '100%',
    }),
    valueContainer: () => ({
        paddingLeft: 56,
    }),
};

const SearchForm: React.FC = () => {
    const dispatch: Dispatch = useDispatch();
    const [selectedValue, setSelectedValue] = useState<OptionTypeBase | null>(
        null
    );
    const isLoading: boolean = useSelector(
        selectRequesting([REQUEST_LOCATIONS])
    );
    const locations: LocationModel[] = useSelector(locationsSelector);

    const onChange = (option: OptionTypeBase | null) => {
        if (option) {
            dispatch(requestLocationWeatherAction(option.value));
            setSelectedValue(option);
        }
    };

    const onInputChange = useCallback(
        (newValue: string) => {
            const inputValue = newValue.replace(/\W/g, '');
            if (inputValue) {
                dispatch(requestLocationsAction(inputValue));
            }
            return inputValue;
        },
        [dispatch]
    );

    const options = useMemo(
        (): ReadonlyArray<OptionTypeBase> =>
            locations.map((location: LocationModel) => ({
                label: location.title,
                value: location.woeid,
            })),
        [locations]
    );

    return (
        <Form className="search-form">
            <InputGroup className="mb-2">
                <InputGroup.Prepend>
                    <InputGroup.Text>üîç</InputGroup.Text>
                </InputGroup.Prepend>
                <Select
                    cacheOptions
                    isLoading={isLoading}
                    noOptionsMessage={() => 'No city'}
                    onChange={onChange}
                    onInputChange={onInputChange}
                    options={options}
                    placeholder="Enter city name"
                    styles={customStyles}
                    value={selectedValue}
                />
            </InputGroup>
        </Form>
    );
};

export default SearchForm;
