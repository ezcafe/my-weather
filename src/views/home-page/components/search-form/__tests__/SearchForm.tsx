import { render, RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SearchForm from '../SearchForm';
import { mockedLocations } from '../../../../../fixtures';
import withMockedRedux from '../../../../../fixtures/withMockedRedux';
import IStore from '../../../../../models/IStore';

describe('SearchForm', () => {
    function renderComponent(initialState?: Partial<IStore>): RenderResult {
        const WrappedComponent = withMockedRedux(SearchForm, initialState);
        return render(<WrappedComponent />);
    }

    test('should show component', () => {
        const { container } = renderComponent();
        expect(container).toMatchSnapshot();
    });

    test('should show loading indicator when typing', () => {
        const { container, getByRole } = renderComponent();
        userEvent.type(getByRole('textbox'), 'san');
        expect(container).toMatchSnapshot();
    });

    test('should show suggestions', () => {
        const { container, getByRole } = renderComponent({
            locations: mockedLocations,
        });
        userEvent.click(getByRole('textbox'));
        expect(container).toMatchSnapshot();
    });
});
