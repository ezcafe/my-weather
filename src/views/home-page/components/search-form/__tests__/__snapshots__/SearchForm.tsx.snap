// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should show component 1`] = `
<div>
  <form
    class="search-form"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          üîç
        </span>
      </div>
      <div
        class=" css-1s1mbqj-SelectContainer"
      >
        <div
          class=" css-yk16xz-control"
        >
          <div
            class=" css-1wcf4nn-ValueContainer"
          >
            <div
              class=" css-1wa3eu0-placeholder"
            >
              Enter city name
            </div>
            <div
              class="css-b8ldur-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="react-select-2-input"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`SearchForm should show loading indicator when typing 1`] = `
<div>
  <form
    class="search-form"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          üîç
        </span>
      </div>
      <div
        class=" css-1s1mbqj-SelectContainer"
      >
        <span
          aria-live="polite"
          class="css-1f43avz-a11yText-A11yText"
        >
          <span
            id="aria-selection-event"
          >
            ¬†
            
          </span>
          <span
            id="aria-context"
          >
            ¬†
              0 results available for search term san. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
          </span>
        </span>
        <div
          class=" css-1pahdxg-control"
        >
          <div
            class=" css-1wcf4nn-ValueContainer"
          >
            <div
              class="css-b8ldur-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="react-select-3-input"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value="san"
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                >
                  san
                </div>
              </div>
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <div
              aria-hidden="true"
              class=" css-1xh8qep-loadingIndicator"
            >
              <span
                class="css-1yvy2vo-LoadingDot"
              />
              <span
                class="css-zoievk-LoadingDot"
              />
              <span
                class="css-x748d8-LoadingDot"
              />
            </div>
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1gtu0rj-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class=" css-26l3qy-menu"
        >
          <div
            class=" css-4ljt47-MenuList"
          >
            <div
              class=" css-1tfbn7p-LoadingMessage"
            >
              Loading...
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`SearchForm should show suggestions 1`] = `
<div>
  <form
    class="search-form"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          üîç
        </span>
      </div>
      <div
        class=" css-1s1mbqj-SelectContainer"
      >
        <span
          aria-live="polite"
          class="css-1f43avz-a11yText-A11yText"
        >
          <span
            id="aria-selection-event"
          >
            ¬†
            
          </span>
          <span
            id="aria-context"
          >
            ¬†
             option San Francisco focused, 1 of 1. 1 result available. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
          </span>
        </span>
        <div
          class=" css-1pahdxg-control"
        >
          <div
            class=" css-1wcf4nn-ValueContainer"
          >
            <div
              class=" css-1wa3eu0-placeholder"
            >
              Enter city name
            </div>
            <div
              class="css-b8ldur-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="react-select-4-input"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1gtu0rj-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class=" css-26l3qy-menu"
        >
          <div
            class=" css-4ljt47-MenuList"
          >
            <div
              class=" css-1n7v3ny-option"
              id="react-select-4-option-0"
              tabindex="-1"
            >
              San Francisco
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
